providers/mlx5/verbs.c

```diff
@@ -325,14 +325,97 @@ static void mlx5_insert_dyn_uuars(struct mlx5_context *ctx,
 		if (!bf_uar->dyn_alloc_uar)
 			bf->bfreg_dyn_index = (ctx->curr_legacy_dyn_sys_uar_page - 1) * num_bfregs_per_page;
 		bf->dyn_alloc_uar = bf_uar->dyn_alloc_uar;
-		bf->need_lock = 0;
+		bf->need_lock = bf_uar->qp_shared;
+		mlx5_spinlock_init(&bf->lock, bf->need_lock);
 		if (j != 0) {
 			bf->uar = bf_uar->uar;
 			bf->page_id = bf_uar->page_id + index_uar_in_page;
 			bf->uar_handle = bf_uar->uar_handle;
 			bf->nc_mode = bf_uar->nc_mode;
 		}
+		if (bf_uar->qp_dedicated) {
+			ctx->qp_alloc_dedicated_uuars++;
+			bf->qp_dedicated = true;
+		} else if (bf_uar->qp_shared) {
+			ctx->qp_alloc_shared_uuars++;
+			bf->qp_shared = true;
+		}
+	}
+}
+
+static void mlx5_put_qp_uar(struct mlx5_context *ctx, struct mlx5_bf *bf)
+{
+	if (!bf || (!bf->qp_dedicated && !bf->qp_shared))
+		return;
+
+	pthread_mutex_lock(&ctx->dyn_bfregs_mutex);
+	if (bf->qp_dedicated)
+		list_add_tail(&ctx->dyn_uar_qp_dedicated_list,
+			      &bf->uar_entry);
+	else
+		bf->count--;
+	pthread_mutex_unlock(&ctx->dyn_bfregs_mutex);
+}
+
+static int mlx5_alloc_qp_uar(struct ibv_context *context, bool dedicated)
+{
+	struct mlx5_context *ctx = to_mctx(context);
+	struct mlx5_bf *bf;
+	uint32_t flags;
+
+	flags = (ctx->shut_up_bf || !ctx->bf_reg_size) ?
+		MLX5_IB_UAPI_UAR_ALLOC_TYPE_NC :
+		MLX5_IB_UAPI_UAR_ALLOC_TYPE_BF;
+
+	bf = mlx5_alloc_dyn_uar(context, flags);
+	if (!bf)
+		return -1;
+
+	if (dedicated)
+		bf->qp_dedicated = true;
+	else
+		bf->qp_shared = true;
+
+	mlx5_insert_dyn_uuars(ctx, bf);
+	return 0;
+}
+
+static struct mlx5_bf *mlx5_get_qp_uar(struct ibv_context *context)
+{
+	struct mlx5_context *ctx = to_mctx(context);
+	struct mlx5_bf *bf = NULL, *bf_entry;
+
+	pthread_mutex_lock(&ctx->dyn_bfregs_mutex);
+
+retry_dedicated:
+	bf = list_pop(&ctx->dyn_uar_qp_dedicated_list, struct mlx5_bf, uar_entry);
+	if (!bf) {
+		if (ctx->qp_alloc_dedicated_uuars < ctx->qp_max_dedicated_uuars) {
+			if (mlx5_alloc_qp_uar(context, true))
+				goto end;
+
+			goto retry_dedicated;
```

Comment:
Use a loop instead of these gotos

Reply:
Done, thanks.
A new series with this change and up-to-date kernel-header patch was uploaded.